File: bridge/PolyBLEPNode.cpp
#include "./dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
#include "struct.h"
using namespace lab;

DART_EXPORT int createPolyBLEPNode(AudioContext* context) {
    auto node = std::make_shared<PolyBLEPNode>(*context);
    return keepNode(node);
}

DART_EXPORT int PolyBLEPNode_type(int nodeId) {
    auto node = std::static_pointer_cast<PolyBLEPNode>(getNode(nodeId));
    return node ? static_cast<int>(node->type()) : 0;
}

DART_EXPORT void PolyBLEPNode_setType(int nodeId, int type) {
    auto node = std::static_pointer_cast<PolyBLEPNode>(getNode(nodeId));
    if(node) node->setType(PolyBLEPType(type));
}

DART_EXPORT int PolyBLEPNode_amplitude(int nodeId) {
    auto node = std::static_pointer_cast<PolyBLEPNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 1, node->amplitude()) : -1;
}

DART_EXPORT int PolyBLEPNode_frequency(int nodeId) {
    auto node = std::static_pointer_cast<PolyBLEPNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 2, node->frequency()) : -1;
}

File: lab_sound_ffi/lib/src/lab_sound/poly_BLEP_node.dart
import 'audio_context.dart';
import 'audio_param.dart';
import 'audio_scheduled_source_node.dart';
import 'lab_sound.dart';

enum PolyBLEPType {
  TRIANGLE,
  SQUARE,
  RECTANGLE,
  SAWTOOTH,
  RAMP,
  MODIFIED_TRIANGLE,
  MODIFIED_SQUARE,
  HALF_WAVE_RECTIFIED_SINE,
  FULL_WAVE_RECTIFIED_SINE,
  TRIANGULAR_PULSE,
  TRAPEZOID_FIXED,
  TRAPEZOID_VARIABLE,
}


class PolyBLEPNode extends AudioScheduledSourceNode {
  late AudioParam amplitude;
  late AudioParam frequency;
  PolyBLEPNode(AudioContext ctx): super(ctx, LabSound().createPolyBLEPNode(ctx.pointer)) {
    amplitude = AudioParam(this.ctx, nodeId, LabSound().PolyBLEPNode_amplitude(nodeId));
    frequency = AudioParam(this.ctx, nodeId, LabSound().PolyBLEPNode_frequency(nodeId));
  }

  PolyBLEPType get type => PolyBLEPType.values[LabSound().PolyBLEPNode_type(nodeId)];
  set type(PolyBLEPType type) => LabSound().PolyBLEPNode_setType(nodeId, type.index);

}

